Fix OSS traces iframe issue

mahmoud-syr opened the pull request:
Related to https://github.com/apaleo/ui/issues/6809

When you work inside an `iframe` you are essentially acting as a third-party. The "new" thing with Apple's Webkit engine and it's `ITP` for detecting trackers and prevent them from tracking users now is that it purges all cookies stored in a third-party context. To fix the issue, the user must first interact with your website as a first-party and you must store your cookies from a first-party context. Afterwards, and once your website got loaded in a third-party context, you must use the `StorageAccessApi` to ask the user for their permission to store cookies.

More details can be found [here](https://developer.mozilla.org/en-US/docs/Web/API/Storage_Access_API), [here](https://webkit.org/blog/8124/introducing-storage-access-api/) and [here](https://webkit.org/blog/7675/intelligent-tracking-prevention/).

What I basically did, is that, I disabled the global authorization middleware for two particular `Razor` pages, namely `_Host` and `Landing`. Added a landing page, once the `_Host` got loaded it checks first for a test/dummy cookie if it presents in the user agent, if not, chances are it's either their first visit ever for our app or they have already visited the app but we are now acting on a third-party context and thus we do not have access to our first party-cookies. So, and in that case, we redirect the user to `Landing` and see if their `user-agent a.k.a browser` requires the requesting a permission to access the storage within a third-party context, if it does, we ask the user for their permission.

That's basically it. 

Test version, not fully functional, missing the S3Storage configuration:

https://mahmoud-traces-app.herokuapp.com

Special thanks to the heroku services for providing this awesome free container host service :D

WinniX commented:
@mahmoud-syr what about other apps then? Like our google spreadsheet addon https://github.com/apaleo/gsheet-addon, or external apps?

mahmoud-syr commented:
> @mahmoud-syr what about other apps then? Like our google spreadsheet addon https://github.com/apaleo/gsheet-addon, or external apps?

On iOS, we are getting `400 Bad Request` when identity, and after a successful login, tries to redirect back to `OSS-traces`. The documentation on `Security-content-policy` mentions that, if you are doing ajax/xmlhttp request in the background to a website that's on a different origin, the request will be denied and the browser will simulate `400 Bad Request` for it. That's why I thought adding identity to the `connect-src` will prevent the issue from happening. But after a discussion with Vasily today, I came to the conclusion that it most probably the cookie and not the `security-content-policy` header.

I reverted the changes for the `security-content-policy` and added `Secure` for the `OSS-traces` app cookie configuration.

So, the most probable scenario now is that, the cookie, and after a successful authentication, didn't get sent back to identity because of the same-site issue. And in order to avoid this issue, the only solution is to set `Samesite=None, Secure` as [this document](https://docs.microsoft.com/en-us/aspnet/core/security/samesite?view=aspnetcore-5.0) suggest.

coherent-edkarhuber commented on src/Traces.Web/Pages/Landing.cshtml:
```diff
@@ -0,0 +1,75 @@
+        <div id="cookie-consent-container">
+            <h4>We need your consent</h4>
+            <p>
+                Your web-browser requires that in order for us to store essential cookeis we need to ask for
```
```suggestion
                Your web-browser requires that in order for us to store essential cookies we need to ask for
```

coherent-edkarhuber commented on src/Traces.Web/Pages/_Host.cshtml.cs:
```diff
@@ -12,12 +13,33 @@ public class HostPageModel : PageModel
+            if (!Request.Cookies.ContainsKey(AppConstants.AppConfigCookieName))
             {
-                AccessToken = await HttpContext.GetTokenAsync(SecurityConstants.AccessToken);
+                var returnUrlWithQuery = Request.Path;
+
+                if (!string.IsNullOrWhiteSpace(Request.QueryString.Value))
+                {
+                    returnUrlWithQuery = $"{returnUrlWithQuery}{Request.QueryString.Value}";
+                }
```
As alternative: Request.GetEncodedPathAndQuery() (from Microsoft.AspNetCore.Http.Extensions)
it's up to you - the result is the same.

ozonni commented on src/Traces.Web/Pages/Landing.cshtml:
```diff
@@ -0,0 +1,75 @@
+</head>
+<body>
+    <div class="main-container">
+        <input id="returnUrlInput" type="hidden" value="@(ViewData["ReturnUrl"])" />
```
you can use `@(ViewData["ReturnUrl"])` in `navigateToMainApp`

ozonni commented on src/Traces.Web/Pages/Landing.cshtml:
```diff
@@ -0,0 +1,75 @@
+    </div>
+
+    <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js" integrity="sha384-UO2eT0CpHqdSJQ6hJty5KVphtPhzWj9WO1clHTMGa3JDZwrnQq4sF86dIHNDz0W1" crossorigin="anonymous"></script>
+    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js" integrity="sha384-JjSmVgyd0p3pXB1rRibZUAYoIIy6OrQ6VrjIEaFf/nJGzIxFDsf4x0xIM+B07jRM" crossorigin="anonymous"></script>
```
check if you really need this

ozonni commented on src/Traces.Web/wwwroot/js/userAgentUtills.js:
```diff
@@ -0,0 +1,67 @@
+        }
+        document.cookie = name + "=" + (value || "") + expires + "; SameSite=None; Secure; path=/";
+    },
+    getCookie: function(name) {
```
not used

ozonni commented on src/Traces.Web/wwwroot/js/userAgentUtills.js:
```diff
@@ -0,0 +1,67 @@
+            try {
+                return document.requestStorageAccess()
+                    .then(() => {
+                        console.log('storage access granted successfully');
```
clean up?

ozonni commented on src/Traces.Web/wwwroot/js/userAgentUtills.js:
```diff
@@ -0,0 +1,53 @@
+        try {
+            return window.self !== window.top;
+        } catch (e) {
+            return true;
```
is it necessary to catch?

ozonni commented on src/Traces.Web/wwwroot/js/userAgentUtills.js:
```diff
@@ -0,0 +1,53 @@
+            return true;
+        }
+    },
+    setCookie: function(name,value,days=1) {
```
```suggestion
    setCookie: function(name, value) {
```

ozonni commented on src/Traces.Web/wwwroot/js/userAgentUtills.js:
```diff
@@ -0,0 +1,53 @@
+    },
+    setCookie: function(name,value,days=1) {
+        var expires = "";
+        if (days) {
```
not needed then, right?

ozonni commented on src/Traces.Web/wwwroot/js/userAgentUtills.js:
```diff
@@ -0,0 +1,53 @@
+    hasStorageAccess: function(){
+        return new Promise((resolve, reject) => {
+            try {
+                if (!userAgentUtills.isUserAgentSupportStorageAccessApi()) {
```
let's move this out since it's not a job of this method

ozonni commented on src/Traces.Web/Pages/Landing.cshtml:
```diff
@@ -0,0 +1,133 @@
+    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css" integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">
+    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.4.1/css/all.css" integrity="sha384-5sAR7xN1Nv6T6+dT2mhtzEpVJvfS3NScPQTrOxhwjIuvcA67KV2R5Jz6kr4abQsz" crossorigin="anonymous">
+    <link href="css/site.css" rel="stylesheet" />
+        <style>
```
let's move it into the file

ozonni commented on src/Traces.Web/Pages/Landing.cshtml:
```diff
@@ -0,0 +1,133 @@
+    </div>
+
+    <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
+    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js" integrity="sha384-JjSmVgyd0p3pXB1rRibZUAYoIIy6OrQ6VrjIEaFf/nJGzIxFDsf4x0xIM+B07jRM" crossorigin="anonymous"></script>
```
do you need boostrap.js?

ozonni commented on src/Traces.Web/Pages/Landing.cshtml:
```diff
@@ -0,0 +1,133 @@
+    <script src="js/userAgentUtills.js"></script>
+    <script type="text/javascript">
+        $(document).ready(function(){
+            userAgentUtills.hasStorageAccess().then(hasStorageAccess => {
```
mb let's do `userAgentUtills.isUserAgentSupportStorageAccessApi()` first before triggering `hasStorageAccess` ?

ozonni commented:
+ the button style on Iphone
